### PR TITLE
Make PEM use cipher.c when available

### LINKED ISSUE #8051: Make PEM use cipher.c when available
Currently our PEM module makes direct calls to `mbedtls_aes` and `mbedtls_des` function. Instead it should use the `mbedtls_cipher` API, for a number of reasons:
- single implementation for both ciphers should save code size when they're both enabled;
- when cipher gains the ability to dispatch to PSA, PEM will automatically benefit from that.

For this issue, the new implementation based on cipher should guarded by `#if defined(MBEDTLS_CIPHER_C)`, and the current implementation kept in the `#else` branch. (Testing notes: the default configuration has `CIPHER_C`, and we have a dedicated component without it, so we're fine already.)

(Note: later, when we have defined `MBEDTLS_CIPHER_LIGHT`, we'll auto-enable `CIPHER_LIGHT` when `PEM_C` is enabled, and remove the implementation based on `mbedtls_aes` and `mbedtls_des`. But no yet.)

Note: the Cipher implementation should explicitly set padding to `MBEDTLS_PADDING_NONE` (which is the only option for the low-level `mbedtls_aes` API, but the Cipher layer supports other options, so we need to select this one explicitly).

### PR DESCRIPTION
**Pull request by valeriosetti**
Resolves #8051

## PR checklist

- [ ] **changelog** since `CIPHER_C` is auto-enabled when `PEM_PARSE_C` is enabled, do we need to track this change anyway?
- [ ] **backport** not required
- [ ] **tests** provided: a couple of new tests were added

### CODE REVIEW

File: library/pem.c [hunk 1]
```diff
@@ -158,99 +158,63 @@ static int pem_pbkdf1(unsigned char *key, size_t keylen,
...
+    memcpy(input, output, input_len);
 
 exit:
-    mbedtls_aes_free(&aes_ctx);
-    mbedtls_platform_zeroize(aes_key, keylen);
-
+    mbedtls_cipher_free(&ctx);
+    if (output != NULL) {
+        mbedtls_free(output);
+    }
```
Comment (gabor-mezei-arm) [inline library/pem.c:214]: The content of the `key` variable must be erased from the memory with `mbedtls_platform_zeroize()`.

File: library/pem.c [hunk 2]
```diff
@@ -158,99 +158,63 @@ static int pem_pbkdf1(unsigned char *key, size_t keylen,
...
-
-    if ((ret = pem_pbkdf1(des_key, 8, des_iv, pwd, pwdlen)) != 0) {
-        goto exit;
-    }
-
-    if ((ret = mbedtls_des_setkey_dec(&des_ctx, des_key)) != 0) {
-        goto exit;
+    mbedtls_cipher_context_t ctx;
+    // Pick the largest key size between DES and AES, i.e. AES-256
+    unsigned char key[32];
```
Comment (gabor-mezei-arm) [inline library/pem.c:169]: I prefer to use the `MBEDTLS_MAX_KEY_LENGTH` instead.

**Review by gabor-mezei-arm — CHANGES_REQUESTED** [review]

**Review by gabor-mezei-arm — APPROVED** [review]
Looks good to me

**Comment by valeriosetti** [comment]
I just rebased this PR in order to solve conflicts with `development`

**Comment by mpg** [comment]
On hold: https://github.com/Mbed-TLS/mbedtls/issues/8051#issuecomment-1759215720

**Comment by mpg** [comment]
Closing as we're not going to address PEM with driver-only ciphers before 4.0, and after 4.0 the strategy will be different (and much simpler).

**Pull request closed without merging**